rs/replicated_state/src/canister_state/queues/message_pool.rs

```diff
@@ -119,8 +123,103 @@ impl Id {
     }
 }
 
-impl From<&Id> for pb_queues::canister_queue::QueueItem {
-    fn from(item: &Id) -> Self {
+/// A typed reference -- inbound (`CanisterInput`) or outbound
+/// (`RequestOrResponse`) -- to a message in the `MessagePool`.
+#[derive(Debug)]
+pub(super) struct Reference<T>(u64, PhantomData<T>);
+
+impl<T> Reference<T> {
+    pub(super) fn kind(&self) -> Kind {
+        Id::from(self).kind()
+    }
+
+    pub(super) fn context(&self) -> Context {
+        Id::from(self).context()
+    }
+
+    pub(super) fn class(&self) -> Class {
+        Id::from(self).class()
+    }
+}
+
+impl<T> Clone for Reference<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T> Copy for Reference<T> {}
```

Comment:
Annotations appear to only work on structs, not traits. Or maybe I'm just missing something.

The compiler won't complain if I add a `#[derive(Copy)]` annotation instead; it will complain if I add both the annotation and the implementation; but if I only have the annotation it will complain elsewhere (where `Copy` would be needed) that `Reference` doesn't implement `Copy`.

So I just went with what seemed to work.

Reply:
See [this](https://stegosaurusdormant.com/understanding-derive-clone/) for more context. Basically `#[derive(Copy)]` ends up generating something like `impl<T> Copy for Reference<T> where T: Copy`. And because neither `CanisterInput` nor `RequestOrResponse` are `Copy`, the attribute does nothing.